File: python_scripts/rra_premium_processing.py
"""
RRA Premium Processing Script for Power BI
Solvency II Pillar 3 - Risk and Claim Reporting
Reserve Return Annual (RRA) Forms Generator

This script processes Lloyd's premium data and generates output tables
required by the RRA - Reserve Return Annual forms.

Compatible with Power BI Python visual and data transformations.
"""

import pandas as pd
import numpy as np
from datetime import datetime

class RRAPremiumProcessor:
    """
    Process premium data for RRA reporting requirements
    """

    def __init__(self, premium_data):
        """
        Initialize processor with premium data

        Parameters:
        -----------
        premium_data : pandas.DataFrame
            DataFrame containing premium data with required columns
        """
        self.premium_data = premium_data
        self.required_columns = [
            'Syndicate Number', 'UMR', 'Insured Country', 'Risk Location',
            'Insured Name', 'Insured - Policyholder Type', 'Risk Code',
            'Risk / Certificate Reference', 'Original Currency', 'Sum Insured',
            'Gross Annual Premium in Period', 'YOA', 'Part VII Indicator'
        ]
        self._validate_data()

    def _validate_data(self):
        """Validate that required columns exist in the data"""
        missing_cols = [col for col in self.required_columns if col not in self.premium_data.columns]
        if missing_cols:
            raise ValueError(f"Missing required columns: {missing_cols}")

    def create_premium_summary_by_syndicate(self):
        """
        Create premium summary aggregated by syndicate

        Returns:
        --------
        pandas.DataFrame : Premium summary by syndicate
        """
        summary = self.premium_data.groupby(['Syndicate Number', 'YOA']).agg({
            'Gross Annual Premium in Period': ['sum', 'mean', 'count'],
            'Sum Insured': 'sum',
            'UMR': 'nunique'
        }).reset_index()

        summary.columns = [
            'Syndicate Number', 'YOA',
            'Total Premium', 'Average Premium', 'Number of Risks',
            'Total Sum Insured', 'Unique UMRs'
        ]

        summary = summary.round(2)
        return summary

    def create_premium_by_risk_code(self):
        """
        Create premium analysis by risk code

        Returns:
        --------
        pandas.DataFrame : Premium breakdown by risk code
        """
        risk_summary = self.premium_data.groupby(['Risk Code', 'YOA', 'Syndicate Number']).agg({
            'Gross Annual Premium in Period': 'sum',
            'Sum Insured': 'sum',
            'UMR': 'count'
        }).reset_index()

        risk_summary.columns = [
            'Risk Code', 'YOA', 'Syndicate Number',
            'Total Premium', 'Total Sum Insured', 'Number of Policies'
        ]

        risk_summary = risk_summary.round(2)
        return risk_summary

    def create_geographic_analysis(self):
        """
        Create geographic analysis of premiums

        Returns:
        --------
        pandas.DataFrame : Premium analysis by country
        """
        geo_summary = self.premium_data.groupby(['Insured Country', 'YOA']).agg({
            'Gross Annual Premium in Period': 'sum',
            'Sum Insured': 'sum',
            'Syndicate Number': 'nunique',
            'UMR': 'count'
        }).reset_index()

        geo_summary.columns = [
            'Country', 'YOA',
            'Total Premium', 'Total Sum Insured',
            'Number of Syndicates', 'Number of Policies'
        ]

        # Calculate market share percentage
        total_by_year = geo_summary.groupby('YOA')['Total Premium'].transform('sum')
        geo_summary['Premium Share %'] = (geo_summary['Total Premium'] / total_by_year * 100).round(2)

        geo_summary = geo_summary.round(2)
        return geo_summary

    def create_policyholder_type_analysis(self):
        """
        Create analysis by policyholder type (Retail vs Business)

        Returns:
        --------
        pandas.DataFrame : Premium analysis by policyholder type
        """
        type_summary = self.premium_data.groupby([
            'Insured - Policyholder Type', 'YOA', 'Syndicate Number'
        ]).agg({
            'Gross Annual Premium in Period': ['sum', 'mean'],
            'Sum Insured': 'sum',
            'UMR': 'count'
        }).reset_index()

        type_summary.columns = [
            'Policyholder Type', 'YOA', 'Syndicate Number',
            'Total Premium', 'Average Premium', 'Total Sum Insured', 'Number of Policies'
        ]

        type_summary = type_summary.round(2)
        return type_summary

    def create_currency_analysis(self):
        """
        Create currency breakdown analysis

        Returns:
        --------
        pandas.DataFrame : Premium analysis by currency
        """
        currency_summary = self.premium_data.groupby(['Original Currency', 'YOA']).agg({
            'Gross Annual Premium in Period': 'sum',
            'Sum Insured': 'sum',
            'UMR': 'count'
        }).reset_index()

        currency_summary.columns = [
            'Currency', 'YOA',
            'Total Premium', 'Total Sum Insured', 'Number of Policies'
        ]

        # Calculate currency distribution percentage
        total_by_year = currency_summary.groupby('YOA')['Number of Policies'].transform('sum')
        currency_summary['Policy Share %'] = (currency_summary['Number of Policies'] / total_by_year * 100).round(2)

        currency_summary = currency_summary.round(2)
        return currency_summary

    def create_part_vii_analysis(self):
        """
        Create Part VII indicator analysis

        Returns:
        --------
        pandas.DataFrame : Analysis of Part VII transfers
        """
        part_vii_summary = self.premium_data.groupby(['Part VII Indicator', 'YOA', 'Syndicate Number']).agg({
            'Gross Annual Premium in Period': 'sum',
            'Sum Insured': 'sum',
            'UMR': 'count'
        }).reset_index()

        part_vii_summary.columns = [
            'Part VII Indicator', 'YOA', 'Syndicate Number',
            'Total Premium', 'Total Sum Insured', 'Number of Policies'
        ]

        part_vii_summary = part_vii_summary.round(2)
        return part_vii_summary

    def create_detailed_risk_register(self):
        """
        Create detailed risk register for RRA reporting

        Returns:
        --------
        pandas.DataFrame : Detailed risk-level data
        """
        risk_register = self.premium_data.copy()

        # Add calculated fields
        risk_register['Premium to Sum Insured Ratio'] = (
            risk_register['Gross Annual Premium in Period'] / risk_register['Sum Insured'] * 100
        ).round(2)

        # Add risk classification based on premium size
        def classify_risk_size(premium):
            if premium < 50000:
                return 'Small'
            elif premium < 500000:
                return 'Medium'
            elif premium < 2000000:
                return 'Large'
            else:
                return 'Very Large'

        risk_register['Risk Size Classification'] = risk_register['Gross Annual Premium in Period'].apply(classify_risk_size)

        return risk_register

    def generate_all_outputs(self):
        """
        Generate all RRA output tables

        Returns:
        --------
        dict : Dictionary containing all output DataFrames
        """
        outputs = {
            'Premium_Summary_By_Syndicate': self.create_premium_summary_by_syndicate(),
            'Premium_By_Risk_Code': self.create_premium_by_risk_code(),
            'Geographic_Analysis': self.create_geographic_analysis(),
            'Policyholder_Type_Analysis': self.create_policyholder_type_analysis(),
            'Currency_Analysis': self.create_currency_analysis(),
            'Part_VII_Analysis': self.create_part_vii_analysis(),
            'Detailed_Risk_Register': self.create_detailed_risk_register()
        }

        return outputs

    def export_to_excel(self, output_path='rra_premium_output.xlsx'):
        """
        Export all outputs to Excel file with multiple sheets

        Parameters:
        -----------
        output_path : str
            Path to output Excel file
        """
        outputs = self.generate_all_outputs()

        with pd.ExcelWriter(output_path, engine='openpyxl') as writer:
            for sheet_name, df in outputs.items():
                df.to_excel(writer, sheet_name=sheet_name, index=False)

        print(f"RRA outputs exported to {output_path}")


def process_premium_data_for_powerbi(dataset):
    """
    Main function for Power BI integration

    This function is designed to be called from Power BI Python visual or script.

    Parameters:
    -----------
    dataset : pandas.DataFrame
        Input dataset from Power BI

    Returns:
    --------
    pandas.DataFrame : Processed output for visualization
    """
    processor = RRAPremiumProcessor(dataset)

    # For Power BI, typically return a single DataFrame for visualization
    # You can modify this to return different outputs based on requirements
    return processor.create_premium_summary_by_syndicate()


# Example usage for standalone execution
if __name__ == "__main__":
    # Load synthetic data
    print("Loading premium data...")
    premium_data = pd.read_csv('../synthetic_data/synthetic_lloyds_premium_data.csv')

    print(f"Loaded {len(premium_data)} premium records")

    # Process data
    print("\nProcessing RRA outputs...")
    processor = RRAPremiumProcessor(premium_data)

    # Generate all outputs
    outputs = processor.generate_all_outputs()

    # Display summary of each output
    print("\n" + "="*80)
    print("RRA OUTPUT TABLES GENERATED")
    print("="*80)

    for name, df in outputs.items():
        print(f"\n{name}:")
        print(f"  Rows: {len(df)}")
        print(f"  Columns: {list(df.columns)}")
        print(f"\n  Preview:")
        print(df.head(5).to_string())
        print("-"*80)

    # Export to Excel
    output_file = '../synthetic_data/rra_premium_outputs.xlsx'
    processor.export_to_excel(output_file)
    print(f"\n✓ All outputs exported to {output_file}")

    # For Power BI demonstration
    print("\n" + "="*80)
    print("POWER BI INTEGRATION EXAMPLE")
    print("="*80)
    print("\nTo use in Power BI, import this module and call:")
    print("  result = process_premium_data_for_powerbi(dataset)")
    print("\nOr create custom visualizations using individual methods:")
    print("  processor = RRAPremiumProcessor(dataset)")
    print("  geographic_data = processor.create_geographic_analysis()")

File: r_scripts/rra_premium_processing.R
# RRA Premium Processing Script for Power BI
# Solvency II Pillar 3 - Risk and Claim Reporting
# Reserve Return Annual (RRA) Forms Generator
#
# This script processes Lloyd's premium data and generates output tables
# required by the RRA - Reserve Return Annual forms.
#
# Compatible with Power BI R visual and data transformations.

# Required libraries
required_packages <- c("dplyr", "tidyr", "readr", "writexl", "lubridate")

# Function to check and install packages
install_if_missing <- function(packages) {
  for (pkg in packages) {
    if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
      message(paste("Installing package:", pkg))
      install.packages(pkg, repos = "http://cran.r-project.org", quiet = TRUE)
      library(pkg, character.only = TRUE)
    }
  }
}

# Install and load required packages
install_if_missing(required_packages)

# RRAPremiumProcessor Class (using R6 for OOP approach)
# For simplicity, we'll use functional programming approach in R

#' Validate Premium Data
#'
#' @param premium_data Data frame containing premium data
#' @return Validated data frame or error
validate_premium_data <- function(premium_data) {
  required_columns <- c(
    "Syndicate Number", "UMR", "Insured Country", "Risk Location",
    "Insured Name", "Insured - Policyholder Type", "Risk Code",
    "Risk / Certificate Reference", "Original Currency", "Sum Insured",
    "Gross Annual Premium in Period", "YOA", "Part VII Indicator"
  )

  missing_cols <- setdiff(required_columns, names(premium_data))

  if (length(missing_cols) > 0) {
    stop(paste("Missing required columns:", paste(missing_cols, collapse = ", ")))
  }

  return(premium_data)
}

#' Create Premium Summary By Syndicate
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with premium summary by syndicate
create_premium_summary_by_syndicate <- function(premium_data) {
  summary_df <- premium_data %>%
    group_by(`Syndicate Number`, YOA) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Average Premium` = mean(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Number of Risks` = n(),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Unique UMRs` = n_distinct(UMR),
      .groups = 'drop'
    ) %>%
    mutate(across(where(is.numeric), ~round(., 2)))

  return(summary_df)
}

#' Create Premium Analysis By Risk Code
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with premium breakdown by risk code
create_premium_by_risk_code <- function(premium_data) {
  risk_summary <- premium_data %>%
    group_by(`Risk Code`, YOA, `Syndicate Number`) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Number of Policies` = n(),
      .groups = 'drop'
    ) %>%
    mutate(across(where(is.numeric), ~round(., 2)))

  return(risk_summary)
}

#' Create Geographic Analysis
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with premium analysis by country
create_geographic_analysis <- function(premium_data) {
  geo_summary <- premium_data %>%
    group_by(Country = `Insured Country`, YOA) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Number of Syndicates` = n_distinct(`Syndicate Number`),
      `Number of Policies` = n(),
      .groups = 'drop'
    ) %>%
    group_by(YOA) %>%
    mutate(
      `Premium Share %` = round((`Total Premium` / sum(`Total Premium`)) * 100, 2)
    ) %>%
    ungroup() %>%
    mutate(across(c(`Total Premium`, `Total Sum Insured`), ~round(., 2)))

  return(geo_summary)
}

#' Create Policyholder Type Analysis
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with premium analysis by policyholder type
create_policyholder_type_analysis <- function(premium_data) {
  type_summary <- premium_data %>%
    group_by(
      `Policyholder Type` = `Insured - Policyholder Type`,
      YOA,
      `Syndicate Number`
    ) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Average Premium` = mean(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Number of Policies` = n(),
      .groups = 'drop'
    ) %>%
    mutate(across(where(is.numeric), ~round(., 2)))

  return(type_summary)
}

#' Create Currency Analysis
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with premium analysis by currency
create_currency_analysis <- function(premium_data) {
  currency_summary <- premium_data %>%
    group_by(Currency = `Original Currency`, YOA) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Number of Policies` = n(),
      .groups = 'drop'
    ) %>%
    group_by(YOA) %>%
    mutate(
      `Policy Share %` = round((`Number of Policies` / sum(`Number of Policies`)) * 100, 2)
    ) %>%
    ungroup() %>%
    mutate(across(c(`Total Premium`, `Total Sum Insured`), ~round(., 2)))

  return(currency_summary)
}

#' Create Part VII Analysis
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with analysis of Part VII transfers
create_part_vii_analysis <- function(premium_data) {
  part_vii_summary <- premium_data %>%
    group_by(`Part VII Indicator`, YOA, `Syndicate Number`) %>%
    summarise(
      `Total Premium` = sum(`Gross Annual Premium in Period`, na.rm = TRUE),
      `Total Sum Insured` = sum(`Sum Insured`, na.rm = TRUE),
      `Number of Policies` = n(),
      .groups = 'drop'
    ) %>%
    mutate(across(where(is.numeric), ~round(., 2)))

  return(part_vii_summary)
}

#' Create Detailed Risk Register
#'
#' @param premium_data Data frame containing premium data
#' @return Data frame with detailed risk-level data
create_detailed_risk_register <- function(premium_data) {
  risk_register <- premium_data %>%
    mutate(
      `Premium to Sum Insured Ratio` = round(
        (`Gross Annual Premium in Period` / `Sum Insured`) * 100, 2
      ),
      `Risk Size Classification` = case_when(
        `Gross Annual Premium in Period` < 50000 ~ "Small",
        `Gross Annual Premium in Period` < 500000 ~ "Medium",
        `Gross Annual Premium in Period` < 2000000 ~ "Large",
        TRUE ~ "Very Large"
      )
    )

  return(risk_register)
}

#' Generate All RRA Outputs
#'
#' @param premium_data Data frame containing premium data
#' @return List of data frames containing all outputs
generate_all_rra_outputs <- function(premium_data) {
  # Validate data first
  premium_data <- validate_premium_data(premium_data)

  outputs <- list(
    Premium_Summary_By_Syndicate = create_premium_summary_by_syndicate(premium_data),
    Premium_By_Risk_Code = create_premium_by_risk_code(premium_data),
    Geographic_Analysis = create_geographic_analysis(premium_data),
    Policyholder_Type_Analysis = create_policyholder_type_analysis(premium_data),
    Currency_Analysis = create_currency_analysis(premium_data),
    Part_VII_Analysis = create_part_vii_analysis(premium_data),
    Detailed_Risk_Register = create_detailed_risk_register(premium_data)
  )

  return(outputs)
}

#' Export RRA Outputs to Excel
#'
#' @param premium_data Data frame containing premium data
#' @param output_path Path to output Excel file
export_rra_outputs <- function(premium_data, output_path = "rra_premium_output.xlsx") {
  outputs <- generate_all_rra_outputs(premium_data)

  # Export to Excel using writexl
  write_xlsx(outputs, output_path)

  message(paste("RRA outputs exported to", output_path))
  return(invisible(NULL))
}

#' Process Premium Data for Power BI
#'
#' Main function for Power BI integration.
#' This function is designed to be called from Power BI R visual or script.
#'
#' @param dataset Data frame from Power BI
#' @return Data frame with processed output for visualization
process_premium_data_for_powerbi <- function(dataset) {
  # Validate data
  dataset <- validate_premium_data(dataset)

  # For Power BI, typically return a single DataFrame for visualization
  # You can modify this to return different outputs based on requirements
  result <- create_premium_summary_by_syndicate(dataset)

  return(result)
}

# Main execution block (for standalone testing)
if (interactive() || !exists("dataset")) {

  main <- function() {
    # Load synthetic data
    message("Loading premium data...")

    # Try to read the CSV file
    data_path <- "../synthetic_data/synthetic_lloyds_premium_data.csv"

    if (!file.exists(data_path)) {
      stop(paste("Data file not found:", data_path))
    }

    premium_data <- read_csv(data_path, show_col_types = FALSE)

    message(paste("Loaded", nrow(premium_data), "premium records"))

    # Process data
    message("\nProcessing RRA outputs...")
    outputs <- generate_all_rra_outputs(premium_data)

    # Display summary of each output
    message("\n", paste(rep("=", 80), collapse = ""))
    message("RRA OUTPUT TABLES GENERATED")
    message(paste(rep("=", 80), collapse = ""))

    for (name in names(outputs)) {
      df <- outputs[[name]]
      message("\n", name, ":")
      message("  Rows: ", nrow(df))
      message("  Columns: ", paste(names(df), collapse = ", "))
      message("\n  Preview:")
      print(head(df, 5))
      message(paste(rep("-", 80), collapse = ""))
    }

    # Export to Excel
    output_file <- "../synthetic_data/rra_premium_outputs_r.xlsx"
    export_rra_outputs(premium_data, output_file)
    message("\n✓ All outputs exported to ", output_file)

    # Power BI integration example
    message("\n", paste(rep("=", 80), collapse = ""))
    message("POWER BI INTEGRATION EXAMPLE")
    message(paste(rep("=", 80), collapse = ""))
    message("\nTo use in Power BI, source this script and call:")
    message("  result <- process_premium_data_for_powerbi(dataset)")
    message("\nOr create custom visualizations using individual functions:")
    message("  geographic_data <- create_geographic_analysis(dataset)")

    return(invisible(outputs))
  }

  # Run main function
  tryCatch({
    main()
  }, error = function(e) {
    message("Error: ", e$message)
  })
}

# For Power BI: When using this script in Power BI, the 'dataset' variable
# will be automatically provided by Power BI. The script will execute and
# return the result for visualization.
if (exists("dataset")) {
  # Power BI execution path
  output <- process_premium_data_for_powerbi(dataset)
  print(output)
}
